### ISSUE TITLE
Run baselines on TREC Core 17

### ISSUE THREAD

**Issue by lintool** [castorini/Anserini#217]
From what I can tell, per #201 Anserini can index NYTimes collection as part of TREC core. This code was committed before we had qrels, so now we should go back and run experiments to generate baselines (QL, BM25, RM3).

**Comment by lintool** [comment]
Topics and qrels are here: https://trec.nist.gov/data/core2017.html

**Comment by lintool** [comment]
@rosequ 

The code doesn't actually run. On `tuna`:

```
nohup sh target/appassembler/bin/IndexCollection -collection TrecCoreCollection \
 -input /tuna1/collections/newswire/NYTcorpus/ -generator JsoupGenerator \
 -index lucene-index.core17_nyt.pos+docvectors+rawdocs -threads 16 \
 -storePositions -storeDocvectors -storeRawDocs -optimize \
 >& log.core17_nyt.pos+docvectors+rawdocs &
```

Log:

```
$ cat log.core17_nyt.pos+docvectors+rawdocs 
nohup: ignoring input
2018-06-07 13:03:46,492 INFO  [main] index.IndexCollection (IndexCollection.java:180) - Collection path: /tuna1/collections/newswire/NYTcorpus/
2018-06-07 13:03:46,493 INFO  [main] index.IndexCollection (IndexCollection.java:181) - Index path: lucene-index.core17_nyt.pos+docvectors+rawdocs
2018-06-07 13:03:46,493 INFO  [main] index.IndexCollection (IndexCollection.java:182) - CollectionClass: TrecCoreCollection
2018-06-07 13:03:46,493 INFO  [main] index.IndexCollection (IndexCollection.java:183) - Generator: JsoupGenerator
2018-06-07 13:03:46,494 INFO  [main] index.IndexCollection (IndexCollection.java:184) - Threads: 16
2018-06-07 13:03:46,494 INFO  [main] index.IndexCollection (IndexCollection.java:185) - Keep stopwords? false
2018-06-07 13:03:46,494 INFO  [main] index.IndexCollection (IndexCollection.java:186) - Store positions? true
2018-06-07 13:03:46,494 INFO  [main] index.IndexCollection (IndexCollection.java:187) - Store docvectors? true
2018-06-07 13:03:46,494 INFO  [main] index.IndexCollection (IndexCollection.java:188) - Store transformed docs? false
2018-06-07 13:03:46,494 INFO  [main] index.IndexCollection (IndexCollection.java:189) - Store raw docs? true
2018-06-07 13:03:46,494 INFO  [main] index.IndexCollection (IndexCollection.java:190) - Optimize (merge segments)? true
2018-06-07 13:03:46,502 INFO  [main] index.IndexCollection (IndexCollection.java:214) - Starting indexer...
2018-06-07 13:03:46,712 INFO  [main] index.IndexCollection (IndexCollection.java:236) - 1 files found in /tuna1/collections/newswire/NYTcorpus
2018-06-07 13:03:46,748 ERROR [pool-3-thread-1] generator.LuceneDocumentGenerator (LuceneDocumentGenerator.java:103) - Error extracting document text, skipping document: 0
java.lang.IllegalArgumentException: String input must not be null
	at org.jsoup.helper.Validate.notNull(Validate.java:26) ~[jsoup-1.8.3.jar:?]
	at org.jsoup.parser.TreeBuilder.initialiseParse(TreeBuilder.java:26) ~[jsoup-1.8.3.jar:?]
	at org.jsoup.parser.TreeBuilder.parse(TreeBuilder.java:42) ~[jsoup-1.8.3.jar:?]
	at org.jsoup.parser.HtmlTreeBuilder.parse(HtmlTreeBuilder.java:53) ~[jsoup-1.8.3.jar:?]
	at org.jsoup.parser.Parser.parse(Parser.java:89) ~[jsoup-1.8.3.jar:?]
	at org.jsoup.Jsoup.parse(Jsoup.java:58) ~[jsoup-1.8.3.jar:?]
	at io.anserini.index.transform.JsoupStringTransform.apply(JsoupStringTransform.java:27) ~[anserini-0.0.1-SNAPSHOT.jar:?]
	at io.anserini.index.transform.JsoupStringTransform.apply(JsoupStringTransform.java:24) ~[anserini-0.0.1-SNAPSHOT.jar:?]
	at io.anserini.index.generator.LuceneDocumentGenerator.createDocument(LuceneDocumentGenerator.java:101) [anserini-0.0.1-SNAPSHOT.jar:?]
	at io.anserini.index.IndexCollection$IndexerThread.run(IndexCollection.java:152) [anserini-0.0.1-SNAPSHOT.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]
2018-06-07 13:03:46,753 INFO  [pool-3-thread-1] index.IndexCollection$IndexerThread (IndexCollection.java:160) - tools/build.xml: 0 docs added.
2018-06-07 13:03:46,807 INFO  [main] index.IndexCollection (IndexCollection.java:277) - Indexed documents: 0
2018-06-07 13:03:46,807 INFO  [main] index.IndexCollection (IndexCollection.java:278) - Empty documents: 0
2018-06-07 13:03:46,807 INFO  [main] index.IndexCollection (IndexCollection.java:279) - Errors: 1
2018-06-07 13:03:46,813 INFO  [main] index.IndexCollection (IndexCollection.java:282) - Total 0 documents indexed in 00:00:00
```

Doesn't seem to be excluding files it's supposed to?

**Comment by rosequ** [comment]
@lintool are these text files? They are supposed to be text files. I don't have access to tuna. It runs fine on Hops. I can get access to `tuna` and have a look.

**Comment by Kytabyte** [comment]
@lintool I've done this at my place. I've added the test topics and qrels into `src/main/resources/topics-and-qrels`, and updated `docs/experiments-core17.md` by using these topics and qrels and adding baseline results. I'll do the pull request soon.

**Comment by Kytabyte** [comment]
I manually decompressed the tarball file for NYT collection to get a set of .xml file then I can get the corpus indexed. Otherwise the code could run but nothing will be retrieved, e.g. if I run

```bash
nohup sh target/appassembler/bin/IndexCollection -collection TrecCoreCollection \
 -input /tuna1/collections/newswire/NYTcorpus/data -generator JsoupGenerator \
 -index lucene-index.core17_nyt.pos+docvectors+rawdocs -threads 16 \
 -storePositions -storeDocvectors -storeRawDocs -optimize \
 >& log.core17_nyt.pos+docvectors+rawdocs &
```

The log will be:

```
2018-06-09 18:48:33,182 INFO  [main] index.IndexCollection (IndexCollection.java:189) - Store docvectors? true
2018-06-09 18:48:33,183 INFO  [main] index.IndexCollection (IndexCollection.java:190) - Store transformed docs? false
2018-06-09 18:48:33,183 INFO  [main] index.IndexCollection (IndexCollection.java:191) - Store raw docs? true
2018-06-09 18:48:33,183 INFO  [main] index.IndexCollection (IndexCollection.java:192) - Optimize (merge segments)? true
2018-06-09 18:48:33,191 INFO  [main] index.IndexCollection (IndexCollection.java:216) - Starting indexer...
2018-06-09 18:48:33,395 INFO  [main] index.IndexCollection (IndexCollection.java:238) - 0 files found in /tuna1/collections/newswire/NYTcorpus/data
2018-06-09 18:48:33,447 INFO  [main] index.IndexCollection (IndexCollection.java:279) - Indexed documents: 0
2018-06-09 18:48:33,447 INFO  [main] index.IndexCollection (IndexCollection.java:280) - Empty documents: 0
2018-06-09 18:48:33,448 INFO  [main] index.IndexCollection (IndexCollection.java:281) - Errors: 0
2018-06-09 18:48:33,453 INFO  [main] index.IndexCollection (IndexCollection.java:284) - Total 0 documents indexed in 00:00:00
```

since no file is read in.

**Comment by lintool** [comment]
For `TrecCollection`, we can index compressed files directly, see: https://github.com/castorini/Anserini/blob/master/src/main/java/io/anserini/collection/TrecCollection.java#L43

Can you port this over to NYT so that we don't need to first decompress.

Decompression leads to multiple copies of the collection lying around, which is bad data management practices...

**Comment by Kytabyte** [comment]
`.gz` or other compressed files in `TrecCollection` are *compressed single file*, while for NYT, the files are in `tgz`, which is *a package of compressed file*. As such, multiple files (or folders) will be in one `.tgz` file while we are using the reader to open it. 

From what I can tell, no other collection feeds a package of data into `Collection.FileSegment` in Anserini. And this actually breaks our API, as a `Collection.FileSegment` class essentially should be an iterator of files that can be read directly. I need to look into the code to see if it is possible to feed the tarball file under current API.

**Comment by lintool** [comment]
Code merged into repo as part of 33a4355249ceb559c1a9dd2231397cc96205da1c
